src/Bugsnag/Middleware.cs

```diff
@@ -91,18 +93,29 @@ public static class InternalMiddleware
     };
 
     /// <summary>
-    /// Strips exceptions from the report if they include any 'ignored classes'
+    /// Ignore the report if any of the exceptions in it are included in the
+    /// IgnoreClasses.
     /// </summary>
-    public static Middleware RemoveIgnoredExceptions = report =>
+    public static Middleware CheckIgnoreClasses = report =>
     {
-      if (report.Configuration.IgnoreClasses != null && report.Configuration.IgnoreClasses.Any())
+      if (!report.Ignored && report.Configuration.IgnoreClasses != null && report.Configuration.IgnoreClasses.Any())
       {
         foreach (var @event in report.Events)
         {
-          @event.Exceptions = @event.Exceptions
-            .Where(e => !report.Configuration.IgnoreClasses.Any(@class => @class == e.ErrorClass))
-            .ToArray();
-          // TODO: if we filter out all of the exceptions should we still send the report?
+          foreach (var ignoredClass in report.Configuration.IgnoreClasses)
+          {
+            foreach (var exception in @event.Exceptions)
+            {
+              if (ignoredClass.IsInstanceOfType(exception.OriginalException))
+              {
+                report.Ignore();
+              }
+
+              if (report.Ignored) break;
+            }
+            if (report.Ignored) break;
+          }
+          if (report.Ignored) break;
```

Comment:
Is there an easy way of reducing the number of `if (report.Ignored) break;` here?

Reply:
yeh we could use `goto` instead

Reply:
When can there be multiple events in a payload? Ruby doesn't have this so might not be a good version to copy. We shouldn't drop other events if there is one ignored event by error class. Usually multiple events are entirely disconnected.